File: patterns/distributed/sharding.py
"""
Sharding Pattern - Python

Partitions data across multiple databases/servers to improve
performance and scalability.
"""


class Shard:
    def __init__(self, id):
        self.id = id
        self.data = {}
    
    def get(self, key):
        return self.data.get(key)
    
    def set(self, key, value):
        self.data[key] = value
    
    def delete(self, key):
        return self.data.pop(key, None)


class ShardManager:
    def __init__(self, num_shards=4):
        self.shards = [Shard(i) for i in range(num_shards)]
    
    def get_shard(self, key):
        # Simple hash-based sharding
        hash_val = hash(key)
        return self.shards[hash_val % len(self.shards)]
    
    def set(self, key, value):
        shard = self.get_shard(key)
        shard.set(key, value)
        print(f"Stored {key} in shard {shard.id}")
    
    def get(self, key):
        shard = self.get_shard(key)
        return shard.get(key)


# Example usage
if __name__ == "__main__":
    print("=== Sharding Pattern ===\n")
    
    shard_manager = ShardManager(4)
    
    shard_manager.set('user1', {'name': 'Alice', 'age': 30})
    shard_manager.set('user2', {'name': 'Bob', 'age': 25})
    shard_manager.set('user3', {'name': 'Charlie', 'age': 35})
    
    print("Retrieved:", shard_manager.get('user1'))

File: patterns/distributed/sharding.js
/**
 * Sharding Pattern - JavaScript
 * 
 * Partitions data across multiple databases/servers to improve
 * performance and scalability.
 */

class Shard {
    constructor(id) {
        this.id = id;
        this.data = new Map();
    }
    
    get(key) {
        return this.data.get(key);
    }
    
    set(key, value) {
        this.data.set(key, value);
    }
    
    delete(key) {
        return this.data.delete(key);
    }
}

class ShardManager {
    constructor(numShards = 4) {
        this.shards = Array.from({ length: numShards }, (_, i) => new Shard(i));
    }
    
    getShard(key) {
        // Simple hash-based sharding
        const hash = this.hash(key);
        return this.shards[hash % this.shards.length];
    }
    
    hash(key) {
        let hash = 0;
        for (let i = 0; i < key.length; i++) {
            hash = ((hash << 5) - hash) + key.charCodeAt(i);
            hash = hash & hash; // Convert to 32-bit integer
        }
        return Math.abs(hash);
    }
    
    set(key, value) {
        const shard = this.getShard(key);
        shard.set(key, value);
        console.log(`Stored ${key} in shard ${shard.id}`);
    }
    
    get(key) {
        const shard = this.getShard(key);
        return shard.get(key);
    }
}

// Example usage
console.log('=== Sharding Pattern ===\n');

const shardManager = new ShardManager(4);

shardManager.set('user1', { name: 'Alice', age: 30 });
shardManager.set('user2', { name: 'Bob', age: 25 });
shardManager.set('user3', { name: 'Charlie', age: 35 });

console.log('Retrieved:', shardManager.get('user1'));

module.exports = { Shard, ShardManager };

